File: session-06-capstone/python/tracing/collector.py
import threading


class TraceCollector:
    """Thread-safe list that records every step a request takes through the system."""

    def __init__(self):
        self._trace: list[str] = []
        self._lock = threading.Lock()

    def add(self, step: str):
        with self._lock:
            self._trace.append(step)

    def get_trace(self) -> list[str]:
        with self._lock:
            return list(self._trace)

    def clear(self):
        with self._lock:
            self._trace.clear()

File: session-06-capstone/go/tracing/collector.go
package tracing

import "sync"

// TraceCollector records every step a request takes through the system.
// Thread-safe via mutex. Only used in tests.
type TraceCollector struct {
	mu    sync.Mutex
	trace []string
}

func NewTraceCollector() *TraceCollector {
	return &TraceCollector{}
}

func (c *TraceCollector) Add(step string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.trace = append(c.trace, step)
}

func (c *TraceCollector) GetTrace() []string {
	c.mu.Lock()
	defer c.mu.Unlock()
	result := make([]string, len(c.trace))
	copy(result, c.trace)
	return result
}

func (c *TraceCollector) Clear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.trace = nil
}
